### PR TITLE
Added single device training support

### PR DESCRIPTION
**Pull request by teng-li**
This will be useful when we do multi-process distributed training.

### CODE REVIEW

**Comment by sammiie** [comment]
To address the "Added single device training support" issue #306, you can modify your code to support training on a single device (e.g., a single GPU or CPU) instead of multiple devices (e.g., multi-GPU training). Here's a general outline of how to achieve this:

Import Required Libraries:
Start by importing the necessary libraries and modules relevant to your project. Ensure you have PyTorch (or another deep learning framework) installed.

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
Define Your Model:
Define your neural network model as you normally would. For single-device training, you don't need to worry about distributed data parallelism.
class YourModel(nn.Module):
    def __init__(self, input_size, output_size):
        super(YourModel, self).__init__()
        # Define your model layers here
        self.fc1 = nn.Linear(input_size, 64)
        self.fc2 = nn.Linear(64, output_size)

    def forward(self, x):
        # Implement the forward pass
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x
Load and Preprocess Data:
Load and preprocess your dataset. You can use PyTorch's DataLoader to create batches of data.
# Load and preprocess your data, create DataLoader
dataset = YourDataset(...)
dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)
Define Loss Function and Optimizer:
Define your loss function and optimizer.
model = YourModel(input_size, output_size).to(device)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
Training Loop:
Implement your training loop, making sure to move your model and data to the target device (CPU or GPU).
for epoch in range(num_epochs):
    for inputs, labels in dataloader:
        inputs, labels = inputs.to(device), labels.to(device)

        optimizer.zero_grad()

        # Forward pass
        outputs = model(inputs)
        loss = criterion(outputs, labels)

        # Backward pass and optimization
        loss.backward()
        optimizer.step()

    print(f'Epoch [{epoch + 1}/{num_epochs}], Loss: {loss.item():.4f}')
Evaluation and Testing:
Implement code to evaluate your trained model on validation and test datasets. You can also compute metrics of interest.
# Evaluation and testing code here
Hyperparameter Tuning:
Fine-tune your hyperparameters based on the specific task and dataset you're working with.

Documentation:
Ensure your code is well-documented with comments to make it clear and understandable.

By following these steps, you can modify your code to support single-device training. This will allow you to train your model efficiently on a single GPU or CPU without the need for multi-device training.
